units/patch_intent.go

```diff
@@ -648,109 +648,182 @@ func (j *patchIntentProcessor) buildTasksAndVariants(patchDoc *patch.Patch, proj
+		// we only need to add dependencies and task group tasks for failed tasks because otherwise
+		// we can rely on them being there from the previous patch.
+		if failedOnly {
+			failedPlusNeeded, err := AddTasksNeededByFailedForReuse(failedTasks, failedTaskDisplayNames, project, vt)
+			if err != nil {
+				return errors.Wrap(err, "getting dependencies for activated tasks")
+			}
+			allFailedPlusNeededTasks = append(allFailedPlusNeededTasks, failedPlusNeeded...)
+			filteredTasks = allFailedPlusNeededTasks
 		}
-		failedTasks = append(failedTasks, tasks...)
+
+		variantTask := vt
+		variantTask.Tasks = utility.StringSliceIntersection(filteredTasks, vt.Tasks)
+
+		if len(variantTask.Tasks) != 0 || len(variantTask.DisplayTasks) != 0 {
+			filteredVariantTasks = append(filteredVariantTasks, variantTask)
+			patchDoc.BuildVariants = append(patchDoc.BuildVariants, vt.Variant)
+		}
+
 	}
 
-	patchDoc.Tasks = failedTasks
+	if failedOnly {
+		patchDoc.Tasks = allFailedPlusNeededTasks
+	}
+	patchDoc.VariantsTasks = filteredVariantTasks
+
 	return nil
 }
 
+func filterToActiveForReuse(reusePatch *patch.Patch, project *model.Project) ([]task.Task, error) {
+	reuseTasks, reuseVariants := getReuseTasksAndVariants(reusePatch, project)
+	query := db.Query(bson.M{
+		task.VersionKey:      reusePatch.Version,
+		task.DisplayNameKey:  bson.M{"$in": reuseTasks},
+		task.ActivatedKey:    true,
+		task.DisplayOnlyKey:  bson.M{"$ne": true},
+		task.BuildVariantKey: bson.M{"$in": reuseVariants},
+	})
+	activatedTasks, err := task.FindAll(query)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting previous patch tasks")
+	}
+
+	return activatedTasks, nil
+
+}
+
+// getReuseTasksAndVariants returns the tasks and variants from VariantsTasks instead of patch.Tasks and patch.BuildVariants
+// Those are used because they already include regex tasks and variants as well as dependencies and task group tasks.
+func getReuseTasksAndVariants(reusePatch *patch.Patch, project *model.Project) ([]string, []string) {
+	reuseTasks := []string{}
+	reuseVariants := []string{}
+
+	for _, vt := range reusePatch.VariantsTasks {
+		tasksInProjectVariant := project.FindTasksForVariant(vt.Variant)
+		for _, t := range vt.Tasks {
+			if utility.StringSliceContains(tasksInProjectVariant, t) {
+				reuseTasks = append(reuseTasks, t)
+			}
+		}
+		reuseVariants = append(reuseVariants, vt.Variant)
+		for _, displayTask := range vt.DisplayTasks {
+			for _, t := range displayTask.ExecTasks {
+				if utility.StringSliceContains(tasksInProjectVariant, t) {
+					reuseTasks = append(reuseTasks, t)
+				}
+			}
+		}
+	}
+
+	return reuseTasks, reuseVariants
+
+}
+
+// AddTasksNeededByFailedForReuse add tasks that failed tasks need to run including dependencies and tasks from single host task groups.
+func AddTasksNeededByFailedForReuse(failedTasks []task.Task, failedTaskDisplayNames []string, project *model.Project, vt patch.VariantTasks) ([]string, error) {
```

Comment:
can includeDependencies be used here instead, if we just convert the failedTasks to tv pairs? 

(Also, i think this function can be private)

Reply:
I made it private. It also adds all tasks in a single host task group, so includeDependencies would only cover part of it.